### ISSUE TITLE
Apparent regression to #62: fsnotify package conflict.

### ISSUE THREAD

**Issue by coventry** [elm-lang/elm-platform#116]
I am running into this issue with elm-lang/elm-reactor@444a57b3ab520ccc9e20499cde84334cb219c446 and elm-lang/elm-package@613c6022cec0ec49687c38f11dde0f76ac9fac9b.  fsnotify 0.2.1 is installed, but I get the same error message as @TeddyDD did.  Nothing changes if I attempt to apply @jvoigtlaender's advice by forcing fsnotify to 0.2.0 in the elm-reactor.cabal file and run cabal install.

```
lap% runhaskell BuildFromSource.hs 0.16
Cloning into 'elm-compiler'...
remote: Counting objects: 22357, done.        
remote: Compressing objects: 100% (49/49), done.        
remote: Total 22357 (delta 15), reused 0 (delta 0), pack-reused 22308        
Receiving objects: 100% (22357/22357), 6.35 MiB | 1.66 MiB/s, done.
Resolving deltas: 100% (13347/13347), done.
Checking connectivity... done.
Cloning into 'elm-package'...
remote: Counting objects: 3367, done.        
remote: Total 3367 (delta 0), reused 0 (delta 0), pack-reused 3367        
Receiving objects: 100% (3367/3367), 618.44 KiB | 858.00 KiB/s, done.
Resolving deltas: 100% (2002/2002), done.
Checking connectivity... done.
Cloning into 'elm-make'...
remote: Counting objects: 955, done.        
remote: Compressing objects: 100% (10/10), done.        
remote: Total 955 (delta 3), reused 0 (delta 0), pack-reused 945        
Receiving objects: 100% (955/955), 216.19 KiB | 0 bytes/s, done.
Resolving deltas: 100% (610/610), done.
Checking connectivity... done.
Cloning into 'elm-reactor'...
remote: Counting objects: 2908, done.        
remote: Compressing objects: 100% (212/212), done.        
remote: Total 2908 (delta 105), reused 0 (delta 0), pack-reused 2695        
Receiving objects: 100% (2908/2908), 1.74 MiB | 1.72 MiB/s, done.
Resolving deltas: 100% (1695/1695), done.
Checking connectivity... done.
Cloning into 'elm-repl'...
remote: Counting objects: 920, done.        
remote: Total 920 (delta 0), reused 0 (delta 0), pack-reused 920        
Receiving objects: 100% (920/920), 159.37 KiB | 0 bytes/s, done.
Resolving deltas: 100% (570/570), done.
Checking connectivity... done.
Downloading the latest package list from hackage.haskell.org
cabal: unrecognised command: sandbox (try --help)
cabal: unrecognised command: sandbox (try --help)
Resolving dependencies...
All the requested packages are already installed:
Use --reinstall if you want to reinstall anyway.
Resolving dependencies...
Configuring elm-make-0.1.2...
Configuring elm-repl-0.4.1...
Building elm-repl-0.4.1...
Building elm-make-0.1.2...
Installed elm-repl-0.4.1
Installed elm-make-0.1.2
Resolving dependencies...
Configuring elm-reactor-0.3.1...
Building elm-reactor-0.3.1...
Failed to install elm-reactor-0.3.1
Last 10 lines of the build log (
/home/coventry/.cabal/logs/elm-reactor-0.3.1.log ):
backend/Socket.hs:21:64:
Couldn't match type `FP.FilePath' with `[Char]'
Expected type: FilePath -> IO ()
Actual type: FP.FilePath -> IO ()
In the return type of a call of `sendHotSwap'
In the fourth argument of `NDevel.treeExtExists', namely
`(sendHotSwap watchedFile connection)'
In a stmt of a 'do' block:
_ <- NDevel.treeExtExists
notifyManager "." "elm" (sendHotSwap watchedFile connection)
cabal: Error: some packages failed to install:
elm-reactor-0.3.1 failed during the building phase. The exception was:
ExitFailure 1
```

**Comment by jvoigtlaender** [comment]
Well, you have a too old version of `cabal`. Look at this line:
```
cabal: unrecognised command: sandbox (try --help)
```
Please read what https://github.com/elm-lang/elm-platform/blob/master/README.md#get-haskell-working has to tell you about versions of GHC and `cabal` you need.

**Comment by coventry** [comment]
@jvoigtlaender, thanks for the pointer.  While the output below is from a repeat run, I get the same error having installed cabal 1.22.4.0, blown away Elm-Package and razed cabal with ghc-pkg-reset.  It is failing with "Not in scope: `<$>'" on line 60 of Elm/Package.hs.  It appears that <$> and <*> are not in the default scome.  If I explicitly import them in the header of Package.hs, it compiles but then I get the same problem in AST/Module/Name.hs.  Do you know what could cause this?

```
lap% runhaskell BuildFromSource.hs 0.16
fatal: destination path 'elm-compiler' already exists and is not an empty directory.
fatal: destination path 'elm-package' already exists and is not an empty directory.
fatal: destination path 'elm-make' already exists and is not an empty directory.
fatal: destination path 'elm-reactor' already exists and is not an empty directory.
fatal: destination path 'elm-repl' already exists and is not an empty directory.
Downloading the latest package list from hackage.haskell.org
Skipping download: Local and remote files match.
Writing a default package environment file to
/home/coventry/src/elm-platform/Elm-Platform/0.16/cabal.sandbox.config
Using an existing sandbox located at
/home/coventry/src/elm-platform/Elm-Platform/0.16/.cabal-sandbox
Resolving dependencies...
In order, the following would be installed:
SHA-1.6.4.2 (via: websockets-0.9.6.1) (new package)
bytestring-mmap-0.2.2 (via: snap-core-0.9.8.0) (new package)
entropy-0.3.7 (via: websockets-0.9.6.1) (new package)
enumerator-0.4.20 (via: websockets-snap-0.9.2.0 snap-server-0.9.5.1 snap-core-0.9.8.0 zlib-enum-0.2.3.1 blaze-builder-enumerator-0.2.1.0 attoparsec-enumerator-0.3.4) (new package)
attoparsec-enumerator-0.3.4 (via: snap-server-0.9.5.1 snap-core-0.9.8.0) (new package)
blaze-builder-enumerator-0.2.1.0 (via: snap-server-0.9.5.1 snap-core-0.9.8.0) (new package)
hinotify-0.3.8.1 (via: fsnotify-0.2.1) (new package)
monads-tf-0.1.0.2 (via: MonadCatchIO-transformers-0.3.1.3) (new package)
MonadCatchIO-transformers-0.3.1.3 (via: snap-server-0.9.5.1 snap-core-0.9.8.0) (new package)
regex-base-0.93.2 (via: regex-posix-0.95.2) (reinstall) (changes: mtl-2.1.2 -> 2.2.1)
regex-posix-0.95.2 (via: snap-core-0.9.8.0) (reinstall)
unix-compat-0.4.1.4 (via: snap-server-0.9.5.1 snap-core-0.9.8.0 fsnotify-0.2.1) (new package)
fsnotify-0.2.1 (new package)
websockets-0.9.6.1 (via: websockets-snap-0.9.2.0) (new package)
zlib-bindings-0.1.1.5 (via: zlib-enum-0.2.3.1) (new package)
zlib-enum-0.2.3.1 (via: snap-core-0.9.8.0) (new package)
snap-core-0.9.8.0 (via: websockets-snap-0.9.2.0 snap-server-0.9.5.1) (new package)
snap-server-0.9.5.1 (via: websockets-snap-0.9.2.0) (new package)
websockets-snap-0.9.2.0 (new package)
cabal: The following packages are likely to be broken by the reinstalls:
regex-compat-0.95.1
Use --force-reinstalls if you want to install anyway.
Resolving dependencies...
Notice: installing into a sandbox located at
/home/coventry/src/elm-platform/Elm-Platform/0.16/.cabal-sandbox
Configuring elm-compiler-0.16...
Building elm-compiler-0.16...
Failed to install elm-compiler-0.16
Build log ( /home/coventry/src/elm-platform/Elm-Platform/0.16/.cabal-sandbox/logs/elm-compiler-0.16.log ):
Configuring elm-compiler-0.16...
Building elm-compiler-0.16...
Preprocessing library elm-compiler-0.16...
[ 9 of 85] Compiling Elm.Package      ( src/Elm/Package.hs, dist/dist-sandbox-4341ad42/build/Elm/Package.o )

src/Elm/Package.hs:60:25:
    Not in scope: `<$>'
    Perhaps you meant `</>' (imported from System.FilePath)

src/Elm/Package.hs:114:16:
    Not in scope: `<$>'
    Perhaps you meant `</>' (imported from System.FilePath)

src/Elm/Package.hs:114:24:
    Not in scope: `<*>'
    Perhaps you meant `</>' (imported from System.FilePath)

src/Elm/Package.hs:210:32:
    Not in scope: `<$>'
    Perhaps you meant `</>' (imported from System.FilePath)

src/Elm/Package.hs:217:19:
    Not in scope: `<$>'
    Perhaps you meant `</>' (imported from System.FilePath)

src/Elm/Package.hs:217:27:
    Not in scope: `<*>'
    Perhaps you meant `</>' (imported from System.FilePath)

src/Elm/Package.hs:217:35:
    Not in scope: `<*>'
    Perhaps you meant `</>' (imported from System.FilePath)
cabal: Error: some packages failed to install:
elm-compiler-0.16 failed during the building phase. The exception was:
ExitFailure 1
elm-make-0.16 depends on elm-compiler-0.16 which failed to install.
elm-package-0.16 depends on elm-compiler-0.16 which failed to install.
elm-repl-0.16 depends on elm-compiler-0.16 which failed to install.
Resolving dependencies...
In order, the following would be installed:
SHA-1.6.4.2 (via: websockets-0.9.6.1) (new package)
bytestring-mmap-0.2.2 (via: snap-core-0.9.8.0) (new package)
elm-compiler-0.16 (via: elm-reactor-0.16) (new package)
entropy-0.3.7 (via: websockets-0.9.6.1) (new package)
enumerator-0.4.20 (via: websockets-snap-0.9.2.0 snap-server-0.9.5.1 snap-core-0.9.8.0 zlib-enum-0.2.3.1 blaze-builder-enumerator-0.2.1.0 attoparsec-enumerator-0.3.4) (new package)
attoparsec-enumerator-0.3.4 (via: snap-server-0.9.5.1 snap-core-0.9.8.0) (new package)
blaze-builder-enumerator-0.2.1.0 (via: snap-server-0.9.5.1 snap-core-0.9.8.0) (new package)
hinotify-0.3.8.1 (via: fsnotify-0.2.1) (new package)
monads-tf-0.1.0.2 (via: MonadCatchIO-transformers-0.3.1.3) (new package)
MonadCatchIO-transformers-0.3.1.3 (via: snap-server-0.9.5.1 snap-core-0.9.8.0) (new package)
regex-base-0.93.2 (via: regex-posix-0.95.2) (reinstall) (changes: mtl-2.1.2 -> 2.2.1)
regex-posix-0.95.2 (via: snap-core-0.9.8.0) (reinstall)
unix-compat-0.4.1.4 (via: snap-server-0.9.5.1 snap-core-0.9.8.0 fsnotify-0.2.1) (new package)
fsnotify-0.2.1 (via: elm-reactor-0.16) (new package)
websockets-0.9.6.1 (via: elm-reactor-0.16 websockets-snap-0.9.2.0) (new package)
zlib-bindings-0.1.1.5 (via: zlib-enum-0.2.3.1) (new package)
zlib-enum-0.2.3.1 (via: snap-core-0.9.8.0) (new package)
snap-core-0.9.8.0 (via: elm-reactor-0.16 websockets-snap-0.9.2.0 snap-server-0.9.5.1) (new package)
snap-server-0.9.5.1 (via: elm-reactor-0.16 websockets-snap-0.9.2.0) (new package)
websockets-snap-0.9.2.0 (via: elm-reactor-0.16) (new package)
elm-reactor-0.16 (new package)
cabal: The following packages are likely to be broken by the reinstalls:
regex-compat-0.95.1
Use --force-reinstalls if you want to install anyway.
lap% 
````

**Comment by jvoigtlaender** [comment]
I asked you to check both GHC and cabal versions. The error messages are telling me that you are not using a GHC from the 7.10 line. Right? The platform page promises that pre-0.16 Elm will compile with GHC 7.8. But not Elm 0.16.

**Comment by coventry** [comment]
Thanks.  Would it make sense for BuildFromSource.hs to check these dependencies?

**Comment by jvoigtlaender** [comment]
If that's possible in Haskell in a platform independent way, then probably yes. I don't know whether that is the case.

**Comment by coventry** [comment]
https://github.com/elm-lang/elm-platform/pull/118

**Comment by jvoigtlaender** [comment]
I guess this issue here can be closed, then? I'll comment on #118 over there.

**Comment by r41d** [comment]
I also ran into this issue when trying to compile `elm-reactor-0.3.1` with `GHC 7.8.4`, `cabal 1.18.1.5` and `elm 0.13`.
For people who come here from googling this error:
I worked around it by running `cabal unpack elm-reactor`, deleting all occurences of "FP." and doing `cabal install` from the elm-reactor-0.3.1 directory.
